### ISSUE TITLE
[Feature Request] Update to Phase Finding Algorithm in testConnectivity.py

### ISSUE THREAD

**Issue by bdorney** [cms-gem-daq-project/vfatqc-python-scripts#248]
## Brief summary of issue

Currently phase set points chosen by `testConnectivity.py` may not be reliable.

### Types of issue

- [ ] Bug report (report an issue with the code)
- [x] Feature request (request for change which adds functionality)

## Expected Behavior

Paraphrasing an email from @evka85:

Phase 15 never works so we should just cut the scan range short (e.g. 0-14).

There should be 2 bad spots in the phase where the communication never works.  However the scan will sometimes only ever find one bad spot due to the finite resolution of the phase shifter.  In this case the bad spot is still there it's just between two values and being near it can cause problems.  The bad phase positions can also vary slightly with temperature by +/-1 position.

The bad phases are expected to be 8 phases apart (7 good values in between bad spots).  But this can vary by +/-1. For this reason the setpoint should never be further than 4 phase counts away from any bad spots.

So lines 417 to 491 should be re-written to find the dead spot(s).  If two dead spots are found then the algorithm should search between them to find the largest "good" window and use the middle phase of that window as the setpoint.  If one dead spot is found the algorithm should search 4 phases in either direction from the dead spot (cyclic boundary conditions) and see which direction yields the largest "good" window and then take the second to the last phase in the appropriate window as the phase set point.

## Current Behavior

The phase finding algorithm in `testConnectivity.py` is:

https://github.com/cms-gem-daq-project/vfatqc-python-scripts/blob/1117886b61e0f2192c89db1313a6f7d74aa69889/testConnectivity.py#L417-L491

Here it loops through the phase scan results and looks to find a number of good phase points in a row.  The phase is considered good if the number of times the phase was good matches the number of attempts it was scanned.  The algorithm will then:

1. search for 5 good phases in a row, selecting the 3rd phase in the set as the phase to assign, 2. if 5 good phases in a row are not found it will search for 4 good phases in a row, selecting the 2nd phase in the set to assign as good,
3. if 4 good phases in a row are not found it will search for 3 good phases in a row, selecting the 2nd phase in a set to assign as good.
4. If 3 good phases in a row are not found it is declared a failure

## Context (for feature requests)

Bad phase setting of VFAT-GBT elinks can cause communication problems with lead to bus errors crashing calibration applications.

## Your Environment

* Version used:
* Shell used:

**Comment by jsturdy** [comment]
@bdorney, is there some sample phase scan data that can be used to test the various cases you're hoping to cover?

**Comment by jsturdy** [comment]
> Phase 15 never works so we should just cut the scan range short (e.g. 0-14).

If you're using cyclic boundary conditions, this probably shouldn't be done, right?
Or, at the very least, the thing that is eventually being checked should make sure that phase 15 is always set to bad.

**Comment by bdorney** [comment]
There's some data from QC7 here:

```
/data/bigdisk/GEM-Data-Taking/GE11_QC8//GE11-X-S-BARI-0011/gbtPhaseScan_GE11-X-S-BARI-0011_2019.04.05.21.46.log
```

And also:

http://cmsonline.cern.ch/cms-elog/1081466

**Comment by jsturdy** [comment]
> There's some data from QC7 here:
> 
> ```
> /data/bigdisk/GEM-Data-Taking/GE11_QC8//GE11-X-S-BARI-0011/gbtPhaseScan_GE11-X-S-BARI-0011_2019.04.05.21.46.log
> ```

For the file (and dropping the `phase == 15` point), is this what you'd expect to extract?
* case 1 (two distinct bad phase points, possible to ignore any wraparound):
```
0 [200. 200.   0. 200. 200. 200. 200. 200. 200. 200.   0. 200. 200. 200. 200.] -> Ideal phase for VFAT0 is:  [6]
2 [200. 200. 200. 200. 200. 200.   0. 200. 200. 200. 200. 200. 200. 200.   0.] -> Ideal phase for VFAT2 is:  [10]
5 [  0. 200. 200. 200. 200. 200. 200. 200.   0. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT5 is:  [4]
8 [200. 200. 200.   0. 200. 200. 200. 200. 200. 200. 200.   0. 200. 200. 200.] -> Ideal phase for VFAT8 is:  [7]
9 [200. 200. 200. 200. 200.   0. 200. 200. 200. 200. 200. 200. 200.   0. 200.] -> Ideal phase for VFAT9 is:  [9]
12 [200.   0. 200. 200. 200. 200. 200. 200. 200.   0. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT12 is:  [5]
15 [200.   0. 200. 200. 200. 200. 200. 200. 200.   0. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT15 is:  [5]
17 [200. 200. 200. 200. 200. 178. 200. 200. 200. 200. 200. 200. 200.  66. 200.] -> Ideal phase for VFAT17 is:  [9]
18 [200. 200. 200. 200.   0. 200. 200. 200. 200. 200. 200. 200. 188. 200. 200.] -> Ideal phase for VFAT18 is:  [8]
22 [  0. 200. 200. 200. 200. 200. 200. 200.   1. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT22 is:  [4]
23 [200. 200. 200.   0. 200. 200. 200. 200. 200. 200. 200.   0. 200. 200. 200.] -> Ideal phase for VFAT23 is:  [7]
```

* case 2 (one distinct bad phase point), here, the behaviour of the wrap at 15 (not shown) matters:
```
1 [200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.   0.] -> Ideal phase for VFAT1 is:  [2, 10]
4 [200.   0. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT4 is:  [5, 13]
6 [200. 200. 200.   0. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT6 is:  [7, 15]
7 [200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.   0. 200. 200. 200.] -> Ideal phase for VFAT7 is:  [15, 7]
10 [200. 200. 200. 200. 200. 200. 200. 200.   0. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT10 is:  [12, 4]
13 [200. 200. 200. 200. 200. 200. 173. 200. 200. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT13 is:  [10, 2]
19 [200. 200.   0. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT19 is:  [6, 14]
20 [200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.   0.] -> Ideal phase for VFAT20 is:  [2, 10]
21 [200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.   0. 200.] -> Ideal phase for VFAT21 is:  [1, 9]
```

* case 3 (zero distinct bad phase points):
```
3 [200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT3 is:  []
11 [200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT11 is:  []
14 [200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT14 is:  []
16 [200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200. 200.] -> Ideal phase for VFAT16 is:  []
```

**Comment by bdorney** [comment]
This looks good; do you have a PR for this or a code snippet?

Also for helping to alleviate Case 2 and Case 3 we could bump the number of phase scans up from 200 to 500 or 1000.  I'm not really sure what the "optimal" number of scans is.

I think one one have to plot optimal phase as a function of N_Scans to understand this (e.g. when does the selected phase no longer change)

**Comment by evka85** [comment]
Also in case 2 and 3 (especially in case 3), instead of or in addition to increasing the number of scans, you could increase the delay between the link_reset and reading of the error counter to give it more time to accumulate errors (the link is checked once per orbit, which is about 90us). I think just increasing this delay would give you more sensitivity and probably would be more time-efficient than running more scans.

**Comment by bdorney** [comment]
@evka85 thanks for this suggestion, right now we have hard coded a wait time of one orbit and we consider the link bad if the sync error count is nonzero:

https://github.com/cms-gem-daq-project/ctp7_modules/blob/develop/src/daq_monitor.cpp#L749-L755

How many orbits do you think we should wait?

**Comment by jsturdy** [comment]
In all cases here, the update I have simply works from the assumption that some scan was done, and the algorithm knows the parameters (`n_scans`).
If better stability is desired, it is expected to be taken care of before the scan is run (unless or until the analysis moves into the CTP7 module, in which case a failed phase determination could trigger a re-do with more points)

> This looks good; do you have a PR for this or a code snippet?

I will put this together, it is currently a standalone analysis based on the text file, so I will have to adapt it to the expected format returned by the phase scan.

**Comment by bdorney** [comment]
> How many orbits do you think we should wait?

I can just increase this arbitrarily to 5 orbits and see how it goes; we can do lab tests to see how it goes.

**Comment by evka85** [comment]
Oh a wait time of just one orbit is way too small in my opinion (it will see obviously bad cases, but not the edge cases).. I would much prefer to just increase it to like 100 or more orbits (maybe even 1000) and not do as many scans (e.g. just drop it to 10 scans).
In fact redoing the scans is basically the same as giving firmware more time to count the errors, it's just that it takes more time because you have to rewrite the registers, so in principle only one or few scans should be ok if enough time is given to the firmware to count the errors (this in theory should give the same results but be more time-efficient).

**Comment by evka85** [comment]
If it really takes a long time to do a scan with reasonable statistics, I could update the firmware to allow user to set the gap between sync checks, so that it could be configured to do more than one check per orbit.

Besides that, if you're currently doing the scan vfat by vfat sequentially, it would probably be more efficient to do them all at once i.e. configure all vfats for a certain phase, then link reset and wait while all of them accumulate statistics, and then read all the error counters, and then move on to the next phase. This way even pausing for 100k orbits for each phase setting should be under half a minute per scan.

**Comment by bdorney** [comment]
Thanks for this feedback @evka. I’ll see if I can rework the SW tomorrow to take this into account.

**Comment by bdorney** [comment]
Hi @evka85 I had a look at the phase scan routine:

https://github.com/cms-gem-daq-project/ctp7_modules/blob/b358ed18fd77ed0625b5f04082df7d19da2d7ff6/src/gbt.cpp#L60-L86

And it seems we are waiting 10ms after a link reset before checking link good, sync error counter, or slow control on the front end.  So I'm not sure if increasing the wait time would help since we are already waiting something like 110 orbits.

**Comment by jsturdy** [comment]
Here's what I put together, it might need further refinement. It ignores 15 as a bad phase, but keeps the value as part of any check for the best phase location:
```python
        for ohN in range(nOHs):
            # Skip masked OH's
            if( not ((args.ohMask >> ohN) & 0x1)):
                continue
            GBT_PHASE_RANGE = 16
            WINDOW = 4
            dict_phases2Save[ohN] = [ 0xdeaddead for x in range(0,24) ]
            for vfat in range(0,24):
                phase2Write = -1
                phaseCounts = np.array([ dict_phaseScanResults[ohN][vfat*GBT_PHASE_RANGE+ph] for ph in range(0,GBT_PHASE_RANGE) ])
                allBadPhases = np.where(phaseCounts!=args.nPhaseScans)[0]
                badPhaseCounts = np.delete(allBadPhases,np.where(allBadPhases==15)[0]) ## remove 15 from the list of bad phases
                phaseSum = 0
                if len(badPhaseCounts) == 1:
                    # only one bad phase point, search forward and backwards for most "good" phases
                    # wraparound needs to be handled by a circular range function
                    # phase == 15 needs to be handled, currently included in the sum
                    frange  = crange(int(badPhaseCounts[0]+1),
                                     int(badPhaseCounts[0])+1+WINDOW,
                                     GBT_PHASE_RANGE)
                    brange  = crange(int(badPhaseCounts[0])-WINDOW,
                                     int(badPhaseCounts[0]),
                                     GBT_PHASE_RANGE)
                    fsum = sum(phaseCounts.take(frange, mode='wrap')) # forward  sum
                    bsum = sum(phaseCounts.take(brange, mode='wrap')) # backward sum
                    if fsum > bsum:
                        phase2Write = (badPhaseCounts[0]+WINDOW)%GBT_PHASE_RANGE
                    elif bsum > fsum
                        phase2Write = (badPhaseCounts[0]-WINDOW)%GBT_PHASE_RANGE
                    else:
                        ## choose the phase that doesn't require passing 15?
                        phase2Write = (badPhaseCounts[0]-WINDOW)%GBT_PHASE_RANGE
                    pass
                elif len(badPhaseCounts) == 2:
                    # exactly two bad phases, pick the midpoint, don't worry about wraparound
                    # shoulw we check the distance?
                    phase2Write = int((badPhaseCounts[1] - badPhaseCounts[0])/2+badPhaseCounts[0])
                    pass
                else:
                    # 0 or more than 2 bad phases, shouldn't happen, how to treat?
                    pass

                if phase2Write > -1:
                    printGreen("Phase {0} will be used for (OH{1},VFAT{2})".format(phase2Write,ohN,vfat))
                    dict_phases2Save[ohN][vfat] = phase2Write
                if dict_phases2Save[ohN][vfat] == 0xdeaddead:
                    listOfBadVFATs.append((ohN,vfat))
                    printRed("I did not find a good phase for (OH{0},VFAT{1})".format(ohN,vfat))
                    failed2FindGoodPhase = True
                    pass
                pass # End loop over VFATs
            pass # End loop over OHs
```

and an example circular range function (done returning a list, rather than as a generator because above a list is needed):
```python
def crange(start,stop,modulus):
    """crange returns a list of the elements in the appropriate range
    x,y where x < y, y < 2*modulo: [ x, x+1 ,,, modulus-1 , 0, ,,, y%modulus ]
    x,y where x > y [ x, x+1 ,,, modulo-1 , 0, ,,, y ]
    all other cases are invalid
    """
    if start < stop:
        return [ x % modulus for x in range(start,stop) ]
    elif stop < start:
        return [ x for x in range(start,modulus) ] + [ x for x in range(0,stop) ]
    else:
        # return []
        raise RangeError("Invalid range specified")
```

**Comment by bdorney** [comment]
So in a real system I have a lot of case 3 behavior still.  Using the default phase scan routine coupled with the phase analysis @jsturdy wrote I obtained:

```
Scanning GBT Phases, this may take a moment please be patient
====================
Phase Scan Results for OH4
====================
  Phase    VFAT0    VFAT1    VFAT2    VFAT3    VFAT4    VFAT5    VFAT6    VFAT7    VFAT8    VFAT9    VFAT10    VFAT11    VFAT12    VFAT13    VFAT14    VFAT15    VFAT16    VFAT17    VFAT18    VFAT19    VFAT20    VFAT21    VFAT22    VFAT23
-------  -------  -------  -------  -------  -------  -------  -------  -------  -------  -------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------
      0      100      100      100       16      100      100      100      100      100      100       100         0       100       100       100       100       100       100       100       100       100       100        20       100
      1      100      100      100      100        0      100      100      100      100      100       100       100         0       100       100         0       100       100       100       100       100       100       100       100
      2       70      100      100      100      100      100      100      100        0      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
      3      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100         0
      4      100        0      100      100      100      100      100      100      100      100       100       100       100       100       100       100         0       100       100       100       100       100       100       100
      5      100      100      100      100      100      100      100      100      100        0       100       100       100       100       100       100       100         0       100       100       100         0       100       100
      6      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
      7      100      100      100      100      100        0      100      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
      8      100      100      100       43      100      100      100      100      100      100         0         0       100       100       100       100       100       100       100       100       100       100         0       100
      9        0      100      100      100        0      100      100      100      100      100       100       100         0       100       100        18       100       100       100         0       100       100       100       100
     10      100      100      100      100      100      100      100      100       20      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
     11      100      100      100      100      100      100      100       74      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100         0
     12      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100        29       100       100       100       100       100       100       100
     13      100      100      100      100      100      100      100      100      100        0       100       100       100         0       100       100       100         0       100       100       100       100       100       100
     14      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
     15        0        0        0        0        0        0        0        0        0        0         0         0         0         0         0         0         0         0         0         0         0         0         0         0
Phase 5 will be used for (OH4,VFAT0)
Phase 0 will be used for (OH4,VFAT1)
I did not find a good phase for (OH4,VFAT2)
Phase 4 will be used for (OH4,VFAT3)
Phase 5 will be used for (OH4,VFAT4)
Phase 3 will be used for (OH4,VFAT5)
I did not find a good phase for (OH4,VFAT6)
Phase 7 will be used for (OH4,VFAT7)
Phase 6 will be used for (OH4,VFAT8)
Phase 9 will be used for (OH4,VFAT9)
Phase 4 will be used for (OH4,VFAT10)
Phase 4 will be used for (OH4,VFAT11)
Phase 5 will be used for (OH4,VFAT12)
Phase 9 will be used for (OH4,VFAT13)
I did not find a good phase for (OH4,VFAT14)
Phase 5 will be used for (OH4,VFAT15)
Phase 8 will be used for (OH4,VFAT16)
Phase 9 will be used for (OH4,VFAT17)
I did not find a good phase for (OH4,VFAT18)
Phase 5 will be used for (OH4,VFAT19)
I did not find a good phase for (OH4,VFAT20)
Phase 1 will be used for (OH4,VFAT21)
Phase 4 will be used for (OH4,VFAT22)
Phase 7 will be used for (OH4,VFAT23)
...
...
```

Then following the suggestion from @evka85 instead of increasing the number of scans I increased the wait time from link reset from 10ms (~110 orbits) to 100ms (~1100 orbits):

```
Scanning GBT Phases, this may take a moment please be patient
====================
Phase Scan Results for OH4
====================
  Phase    VFAT0    VFAT1    VFAT2    VFAT3    VFAT4    VFAT5    VFAT6    VFAT7    VFAT8    VFAT9    VFAT10    VFAT11    VFAT12    VFAT13    VFAT14    VFAT15    VFAT16    VFAT17    VFAT18    VFAT19    VFAT20    VFAT21    VFAT22    VFAT23
-------  -------  -------  -------  -------  -------  -------  -------  -------  -------  -------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------
      0      100      100      100       11      100      100      100      100      100      100       100         0       100       100       100       100       100       100       100       100       100       100        28       100
      1      100      100      100      100        0      100      100      100      100      100       100       100         0       100       100         1       100       100       100       100       100       100       100       100
      2       65      100      100      100      100      100      100      100        0      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
      3      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100         0
      4      100        0      100      100      100      100      100      100      100      100       100       100       100       100       100       100         0       100       100       100       100       100       100       100
      5      100      100      100      100      100      100      100      100      100        0       100       100       100       100       100       100       100         0       100       100       100         0       100       100
      6      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
      7      100      100      100      100      100        0      100      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
      8      100      100      100       46      100      100      100      100      100      100         0         0       100       100       100       100       100       100       100       100       100       100         0       100
      9        0      100      100      100        0      100      100      100      100      100       100       100         0       100       100        16       100       100       100         0       100       100       100       100
     10      100      100      100      100      100      100      100      100       24      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
     11      100      100      100      100      100      100      100       73      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100         0
     12      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100        31       100       100       100       100       100       100       100
     13      100      100      100      100      100      100      100      100      100        0       100       100       100         0       100       100       100         0       100       100       100       100       100       100
     14      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
     15        0        0        0        0        0        0        0        0        0        0         0         0         0         0         0         0         0         0         0         0         0         0         0         0
Phase 5 will be used for (OH4,VFAT0)
Phase 0 will be used for (OH4,VFAT1)
I did not find a good phase for (OH4,VFAT2)
Phase 4 will be used for (OH4,VFAT3)
Phase 5 will be used for (OH4,VFAT4)
Phase 3 will be used for (OH4,VFAT5)
I did not find a good phase for (OH4,VFAT6)
Phase 7 will be used for (OH4,VFAT7)
Phase 6 will be used for (OH4,VFAT8)
Phase 9 will be used for (OH4,VFAT9)
Phase 4 will be used for (OH4,VFAT10)
Phase 4 will be used for (OH4,VFAT11)
Phase 5 will be used for (OH4,VFAT12)
Phase 9 will be used for (OH4,VFAT13)
I did not find a good phase for (OH4,VFAT14)
Phase 5 will be used for (OH4,VFAT15)
Phase 8 will be used for (OH4,VFAT16)
Phase 9 will be used for (OH4,VFAT17)
I did not find a good phase for (OH4,VFAT18)
Phase 5 will be used for (OH4,VFAT19)
I did not find a good phase for (OH4,VFAT20)
Phase 1 will be used for (OH4,VFAT21)
Phase 4 will be used for (OH4,VFAT22)
Phase 7 will be used for (OH4,VFAT23)
...
...
List of Bad (OH,VFAT) pairs: [(4, 2), (4, 6), (4, 14), (4, 18), (4, 20)]
```

And while the phase scan results slightly differ the case 3 VFATs remain the same.

**Comment by bdorney** [comment]
to add to the above, with 1100 orbit wait time and performing double the number of scans (200) the "case 3" VFATs never leave case 3.

**Comment by bdorney** [comment]
So reverting to a ~110 orbit wait time, using 100 scans and increasing the number of slow control transaction iterations (reading `CFG_RUN`, `HW_ID` and `HW_ID_VER` registers per VFAT with a 20us wait time in between) from 10 to 1000 has only helped to resolve two of the four case 3 VFATs:

```
Scanning GBT Phases, this may take a moment please be patient
====================
Phase Scan Results for OH4
====================
  Phase    VFAT0    VFAT1    VFAT2    VFAT3    VFAT4    VFAT5    VFAT6    VFAT7    VFAT8    VFAT9    VFAT10    VFAT11    VFAT12    VFAT13    VFAT14    VFAT15    VFAT16    VFAT17    VFAT18    VFAT19    VFAT20    VFAT21    VFAT22    VFAT23
-------  -------  -------  -------  -------  -------  -------  -------  -------  -------  -------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  -------
-
      0      100      100      100        0      100      100      100      100      100      100       100         0       100       100       100       100       100       100       100       100       100       100         0       100
      1      100      100      100      100        0      100      100      100      100      100       100       100         0       100       100         0       100       100       100       100       100       100       100       100
      2        0      100      100      100      100      100      100      100        0      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
      3      100      100      100      100      100      100       94      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100         0
      4      100        0      100      100      100      100      100      100      100      100       100       100       100       100       100       100         0       100       100       100       100       100       100       100
      5      100      100      100      100      100      100      100      100      100        0       100       100       100       100       100       100       100         0       100       100       100         0       100       100
      6      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100       100       100       100       100        97       100       100       100
      7      100      100      100      100      100        0      100      100      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
      8      100      100      100        0      100      100      100      100      100      100         0         0       100       100       100       100       100       100       100       100       100       100         0       100
      9        0      100      100      100        0      100      100      100      100      100       100       100         0       100       100         0       100       100       100         0       100       100       100       100
     10      100      100      100      100      100      100      100      100        0      100       100       100       100       100       100       100       100       100       100       100       100       100       100       100
     11      100      100      100      100      100      100       98        0      100      100       100       100       100       100       100       100       100       100       100       100       100       100       100         0
     12      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100         0       100       100       100       100       100       100       100
     13      100      100       69      100      100      100      100      100      100        0       100       100       100         0       100       100       100         0       100       100       100         0       100       100
     14      100      100      100      100      100      100      100      100      100      100       100       100       100       100       100       100       100       100       100       100         0       100       100       100
     15        0        0        0        0        0        0        0        0        0        0         0         0         0         0         0         0         0         0         0         0         0         0         0         0
Phase 5 will be used for (OH4,VFAT0)
Phase 0 will be used for (OH4,VFAT1)
Phase 9 will be used for (OH4,VFAT2)
Phase 4 will be used for (OH4,VFAT3)
Phase 5 will be used for (OH4,VFAT4)
Phase 3 will be used for (OH4,VFAT5)
Phase 7 will be used for (OH4,VFAT6)
Phase 7 will be used for (OH4,VFAT7)
Phase 6 will be used for (OH4,VFAT8)
Phase 9 will be used for (OH4,VFAT9)
Phase 4 will be used for (OH4,VFAT10)
Phase 4 will be used for (OH4,VFAT11)
Phase 5 will be used for (OH4,VFAT12)
Phase 9 will be used for (OH4,VFAT13)
I did not find a good phase for (OH4,VFAT14)
Phase 5 will be used for (OH4,VFAT15)
Phase 8 will be used for (OH4,VFAT16)
Phase 9 will be used for (OH4,VFAT17)
I did not find a good phase for (OH4,VFAT18)
Phase 5 will be used for (OH4,VFAT19)
Phase 10 will be used for (OH4,VFAT20)
Phase 9 will be used for (OH4,VFAT21)
Phase 4 will be used for (OH4,VFAT22)
Phase 7 will be used for (OH4,VFAT23)
Writing Found Phases to frontend
...
...
List of Bad (OH,VFAT) pairs: [(4, 14), (4, 18)]
```

But this increased the run time of a phase scan from a few minutes to ~6.5 hours...so I'm not sure this is really feasible...

**Comment by bdorney** [comment]
So previously our case 3 VFATs were:

> `List of Bad (OH,VFAT) pairs: [(4, 2), (4, 6), (4, 14), (4, 18), (4, 20)]`

And now with this really extended slow control options:

> `List of Bad (OH,VFAT) pairs: [(4, 14), (4, 18)]`

And VFAT14 and VFAT18 look solid...

Of the VFATs that changed (VFATs 2, 6, and 20) it seems:

- VFAT2 has moved to case 2, only one bad phase, with `phase=13` giving 31% of attempts ending in failure,
- VFAT6 has moved to case 1, two bad phases, with `phase=3` and `phase=11` but each giving respectively 6% and 2% of attempts ending in failure.
- VFAT20 has also moved to case 1, two bad phases, with `phase=6` and `phase=14` ending in failure.  
   - Phase at 6 is barely noticeable and this would probably be a case 2 VFAT if statistics were decreased.
   - Now here is interesting, `phase=14` ends in complete failure with 100% of attempts bad (although previous attempts with slow control iterations at 10 did not really show this, so the large number of slow control iterations really helped to rule this phase out).

So I think I will try again and see if I can reduce the number of slow control iterations per scan to have the time more controlled.

But for the case 3 VFATs (14 and 18) I think the algorithm should just guess for them...we could try to have a hard coded lookup table for long and short detectors in which we have a "guess" of bad phases (e.g. from exercise above we can estimate what the bad phases would be +/-1 and just say the guess is not these).  But I wish we had some better algorithm...

**Comment by bdorney** [comment]
@jsturdy @evka85 In general for case 3 what I would propose is the following:

- Generate a list [0,14]
- randomly pick a phase between [0,14] inclusive,
- Write the phase,
- Issue a link reset,
- Using an `RPC` function perform 1e3 (or 1e4) reads of `HW_ID` register

If no bus errors are detected, set this phase as the phase to be used, if bus error is detected remove the phase from the list, e.g. `myListOfPhases.remove(phase)` and repeat the above procedure.

This could allow us to decrease the total number of slow control iterations in the procedure (saves time) and then try to just brute force to confirm that the "randomly chosen phase" is good.

**Comment by jsturdy** [comment]
> @jsturdy @evka85 In general for case 3 what I would propose is the following:
> 
> * Generate a list [0,14]
> * randomly pick a phase between [0,14] inclusive,
> * Write the phase,
> * Issue a link reset,
> * Using an `RPC` function perform 1e3 (or 1e4) reads of `HW_ID` register
> 
> If no bus errors are detected, set this phase as the phase to be used, if bus error is detected remove the phase from the list, e.g. `myListOfPhases.remove(phase)` and repeat the above procedure.

So... repeat the phase scan?  What is different about this routine than the initial phase scan?

> 
> This could allow us to decrease the total number of slow control iterations in the procedure (saves time) and then try to just brute force to confirm that the "randomly chosen phase" is good.

Is the last bullet point not what the scan routine you described earlier is doing (modulo reading a few other registers)?  Or you're saying, go back to the original scan that catches the majority of good phases quickly, and for the specific ones where we can't find any "bad" phases, we execute this additional scan?

**Comment by bdorney** [comment]
>  Or you're saying, go back to the original scan that catches the majority of good phases quickly, and for the specific ones where we can't find any "bad" phases, we execute this additional scan?

Yes that's what I'm saying, do the quick scan that identifies case 1 and case 2 VFATs, and then do the in depth scan procedure for the case 3 VFATs until the first "good" position is fully qualified.

**Comment by bdorney** [comment]
In the end I reverted to a quick GBT scan and then added some slow control iterations (10 total) and incorporated the phase finding algorithm @jsturdy wrote plus added a case for catching 3 bad phases; submitting PR.

**Comment by bdorney** [comment]
Addressed by #251
